js/src/main/scala/org/soundsofscala/instrument/PianoSampler.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 Sounds of Scala
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.soundsofscala.instrument
+
+import cats.effect.IO
+import cats.implicits.*
+import org.scalajs.dom
+import org.scalajs.dom.AudioBuffer
+import org.scalajs.dom.AudioContext
+import org.soundsofscala.models.*
+import scala.annotation.tailrec
+
+final case class PianoSampler(samples: Map[SampleKey, AudioBuffer]) extends Sampler {
+
+//order samples into a list of (frequency, SampleKey, AudioBuffer) tuples
+  private val orderedSamples: List[(Double, SampleKey, AudioBuffer)] =
+    samples.map { case (key, buffer) => (key.frequency, key, buffer) }.toList.sortBy {
+      case (f, key, buffer) => f
+    }
+
+// Find the closest sample to the note we have been asked to play
+  private def closestFrequency(
+      sampleFreqs: List[(Double, SampleKey, AudioBuffer)],
+      target: Double): (Double, SampleKey, AudioBuffer) = {
+    @tailrec
```

Comment:
I would add a comment that this is using binary search, or, perhaps better, change the name `loop` to `binarySearch`.

This helps indicate to the reader that the choice of algorithm is intentional.